### PR TITLE
User constraints

### PR DESCRIPTION
**Pull request by epawlowska**

### CODE REVIEW

File: whylog/constraints/__init__.py [hunk 1]
```diff
@@ -1,52 +1,207 @@
...
+        """
+        I.e:
+        - verify({}, ['comp1', 'comp1', 'comp1']) returns True
+        - verify({}, ['comp1', 'hello', 'comp1']) returns False
+        """
+        # first_group_content = group_contents[0]
+        # for group_content in group_contents:
+        #     if not first_group_content == group_content:
+        #         return False
+        # return True
```
Comment (konefalg) [inline whylog/constraints/__init__.py]: I think that this commented code can be removed.

File: whylog/constraints/__init__.py [hunk 2]
```diff
@@ -1,52 +1,202 @@
...
+        self._check_useless_params(correct_param_names, actual_param_names)
+        self._check_mandatory_params(correct_param_names, actual_param_names)
+        self._check_optional_params(correct_param_names, actual_param_names)
+
+    def _check_useless_params(self, correct_param_names, actual_param_names):
+        if actual_param_names - correct_param_names:
+            raise ConstructorParamsError(self.TYPE, correct_param_names, actual_param_names)
+
+    def _check_mandatory_params(self, correct_param_names, actual_param_names):
+        pass
```
Comment (ppolewicz) [inline whylog/constraints/__init__.py:76]: this is a method in an abstract class and it is here for a purpose. But as an author of subclass, I don't know it. Please add a docstring.

File: whylog/constraints/__init__.py [hunk 3]
```diff
@@ -1,52 +1,202 @@
...
 
-class IdenticalConstraint(AbstractConstraint):
-    def __init__(self, groups, params):
+    def _check_optional_params(self, correct_param_names, actual_param_names):
```
Comment (ppolewicz) [inline whylog/constraints/__init__.py:78]: docstring

File: whylog/constraints/__init__.py [hunk 4]
```diff
@@ -1,52 +1,202 @@
...
+        return None
 
-class TimeConstraint(AbstractConstraint):
-    def __init__(self, groups, param_dict):
+    @abstractproperty
+    def MAX_GROUPS_COUNT(self):
+        return None
+
+    @abstractproperty
+    def PARAMS(self):
```
Comment (ppolewicz) [inline whylog/constraints/__init__.py:29]: document it

File: whylog/constraints/__init__.py [hunk 5]
```diff
@@ -1,52 +1,202 @@
...
 @six.add_metaclass(ABCMeta)
 class AbstractConstraint(object):
-    pass
+    """
+    :param MIN_GROUPS_COUNT: minimal groups count needed to create constraint
+    :param MAX_GROUPS_COUNT: maximal groups count needed to create constraint
+    """
+
+    @abstractproperty
+    def TYPE(self):
```
Comment (ppolewicz) [inline whylog/constraints/__init__.py:18]: document it

File: whylog/constraints/__init__.py [hunk 6]
```diff
@@ -1,52 +1,202 @@
...
+
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = 2
+
+    MIN_DELTA = 'min_delta'
+    MAX_DELTA = 'max_delta'
+
+    PARAMS = sorted([MIN_DELTA, MAX_DELTA])
+
+    def __init__(self, groups=None, param_dict=None, params_checking=True):
```
Comment (ppolewicz) [inline whylog/constraints/__init__.py]: constructor from abstract class can be used here. Please move the docstring to the class level and remove this constructor.

File: whylog/constraints/__init__.py [hunk 7]
```diff
@@ -1,52 +1,202 @@
...
+    """
+
+    TYPE = ConstraintType.IDENTICAL
+
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = None
+
+    PARAMS = []
+
+    def __init__(self, groups=None, param_dict=None, params_checking=True):
```
Comment (ppolewicz) [inline whylog/constraints/__init__.py]: constructor from abstract class can be used here. Please move the docstring to the class level and remove this constructor.

File: whylog/constraints/__init__.py [hunk 8]
```diff
@@ -1,52 +1,202 @@
...
+        super(IdenticalConstraint, self).__init__(groups, param_dict, params_checking)
+
+    def verify(self, group_contents, param_dict=None):
+        """
+        I.e:
+        - verify({}, ['comp1', 'comp1', 'comp1']) returns True
+        - verify({}, ['comp1', 'hello', 'comp1']) returns False
+        """
+
+        if len(group_contents) <= 1:
```
Comment (ppolewicz) [inline whylog/constraints/__init__.py]: remove that `if` block. This condition is already checked by `AbstractConstraint._check_constructor_groups()` once per life of constraint, not once per `verify()` call. This approach is like a billion times more effective.

File: whylog/constraints/exceptions.py [hunk 9]
```diff
@@ -11,3 +15,38 @@ def __init__(self, constraint_data):
...
+        self.maximal_groups_count = maximal_groups_count
+
+    def __str__(self):
+        return 'Wrong groups count in constraint: %s, has %s groups, should be at least: %s, at most %s' % (
+            self.constraint_type, self.groups_count, self.minimal_groups_count,
+            self.maximal_groups_count
+        )
+
+
+class VerificatedParamsError(ConstraintError):
```
Comment (ppolewicz) [inline whylog/constraints/exceptions.py]: nobody uses this. Remove it.

**Pull request merged**